src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java

```diff
@@ -0,0 +1,125 @@
+package com.es.phoneshop.dao.impl;
+
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.exceptions.ProductNotFoundException;
+import com.es.phoneshop.model.product.Product;
+
+import java.math.BigDecimal;
+import java.security.InvalidParameterException;
+import java.util.ArrayList;
+import java.util.Currency;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.IntStream;
+
+
+public class ArrayListProductDao implements ProductDao {
+
+    private ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private Lock readLock = readWriteLock.readLock();
+    private Lock writeLock = readWriteLock.writeLock();
+    private List<Product> products;
+    private long id;
+
+    public ArrayListProductDao() {
+        products = new ArrayList<>();
+        saveDefaultSampleProducts();
+    }
+
+    @Override
+    public Product getProduct(Long id) throws ProductNotFoundException, InvalidParameterException {
+        readLock.lock();
+        try {
+            if (id != null) {
+                Product requiredProduct;
+                requiredProduct = products.stream().
+                        filter((product -> id.equals(product.getId())
+                                && product.getPrice() != null
+                                && product.getStock() > 0)).
+                        findFirst().orElseThrow(() -> new ProductNotFoundException(id));
```

Comment:
Throw exception with message. For instance "Product with id 1 doesn't exist "

Reply:
I have stasic string with exception message in exception class. Is it a bad practice and why?

Reply:
It's better to pass message directly in a place where it was thrown. It offeres better readability.